src/runner.js

```diff
@@ -0,0 +1,375 @@
+
+        this.runs.push(Object.freeze({
+            neededToConfirm: neededToConfirm
+        }));
+
+        const runNumber = this.runs.length;
+
+        functionsToRun.forEach(function (def) {
+            def.status = 'running';
+            def.runNumber = runNumber;
+
+            const promise = MiniPromise.createPromise(function (resolve, reject) {
+                try {
+                    const thisArg = { key: toContextKey(def.key) };
+                    const result = def.handler.call(thisArg, resolve, reject);
+
+                    if (result !== true) {
+                        resolve();
+                    }
+                }
+                catch (err) {
+                    reject(err);
+                }
+            });
+
+            promise
+                .then(function () { return def.status = 'resolved'; })
+                .catch(function (err) {
+                    def.status = 'rejected';
+                    client.contaminate({
+                        reason: 'error-thrown',
+                        details: err.message
+                    });
+                    console.warn('[Evolv]: An error occurred while applying a javascript mutation. ' + err);
+                })
+                .finally(function () {
+                    this.checkForConfirmation(runNumber);
+                }.bind(this));
+        }.bind(this));
+    };
+
+    /**
+     * @return FunctionDef[]
+     * @private
+     */
+    Runner.prototype.getConcurrent = function () {
+        if (this.runLevel === levelMap['immediate']) {
+            return this.functions
+                .filter(function (def) { return def.timing === 'legacy'; })
+                .filter(hasNotRun);
+        }
+        else if (this.runLevel === levelMap['legacy']) {
+            return this.functions
+                .filter(function (def) { return def.timing === 'immediate'; })
+                .filter(hasNotCompleted);
+        }
+        else {
+            return [];
+        }
+    };
+
+    /**
+     * @param {number} runNumber
+     * @returns void
+     * @private
+     */
+    Runner.prototype.checkForConfirmation = function (runNumber) {
+        const run = this.runs[runNumber - 1];
+        const allResolved = this.functions
+            .filter(function (def) {
+                return run.neededToConfirm.indexOf(def.key) !== -1;
+            })
+            .every(function (def) {
+                return def.status === 'resolved';
+            });
+        if (allResolved) {
+            this.confirm();
+        }
+    };
+
+    Runner.prototype.print = function () {
+        console.table(this.functions);
+    };
+
+    /**
+     * @returns void
+     * @private
+     */
+    Runner.prototype.confirm = function () {
+        const client = this.container.client;
+        // Q: Could have been a bug? Only confirming on first occurrence, not subsequent context entries
+        if (!this.confirmed) {
```

Comment:
I think this should be removed -- if it's left in then on an SPA we won't try to confirm a new context if the page isn't reload